apps/indexer/src/api/services/coingecko/coingecko.service.ts

```diff
@@ -2,14 +2,30 @@ import { HTTPException } from "hono/http-exception";
 import {
   CoingeckoHistoricalMarketData,
   CoingeckoHistoricalMarketDataSchema,
+  CoingeckoIdToAssetPlatformId,
   CoingeckoTokenId,
 } from "./types";
 import { DAYS_IN_YEAR } from "@/lib/constants";
+import z from "zod";
+
+const createCoingeckoTokenPriceDataSchema = (
+  tokenContractAddress: string,
+  targetCurrency: string,
+) =>
+  z
+    .object({})
+    .catchall(z.object({}).catchall(z.number()))
+    .refine(
+      (data) => data[tokenContractAddress]?.[targetCurrency] !== undefined,
+      {
+        message: `Unable to extract dataset`,
+      },
+    )
+    .transform((data) => data[tokenContractAddress]![targetCurrency]!);
```

Comment:
not sure if this works, but give it a try

```suggestion
const createCoingeckoTokenPriceDataSchema = (
  tokenContractAddress: string,
  targetCurrency: string,
) =>
  z
    .object({
      [tokenContractAddress]: z.object({
        [targetCurrency]: z.number(),
      }),
    })
```

we should rely on the validation coming from zod itself, not through TS shenannigans.

Reply:
Yeah that looks way better. The one thing bothering me is that I needed to assert not null (`!`) in order to avoid this error:

```
typescript: Object is possibly 'undefined'. [2532]
```

When accessing the properties in the returned object.